### PR TITLE
Rewrite password rules to be parameter-driven

### PR DESCRIPTION
**Pull request by labkey-adam**
#### Rationale
Generalize our password rule implementation to increase flexibility and consistency. No real functional change at this point (other than some slight wording changes to messages), but this will make it easy to add new rules and enable custom rules.

#### Related Pull Requests
* https://github.com/LabKey/testAutomation/pull/1581

### CODE REVIEW

File: api/src/org/labkey/api/security/PasswordRule.java [hunk 1]
```diff
@@ -21,182 +21,262 @@
...
-                return false;
+            return 3;
+        }
+    };
 
-            if (SecurityManager.matchesPreviousPassword(password, user))
-            {
-                if (null != messages)
-                    messages.add("Your password must not match a recently used password.");
+    private static final Pattern LOWER_CASE = Pattern.compile("[A-Z]");
```
Comment (labkey-matthewb) [inline api/src/org/labkey/api/security/PasswordRule.java]: Not Unicode compatible.

\p{javaLowerCase}
\p{javaUpperCase}
\W

File: api/src/org/labkey/api/security/PasswordRule.java [hunk 2]
```diff
@@ -21,182 +21,262 @@
...
+import java.util.List;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
-/**
- * User: adam
- * Date: Jan 13, 2010
- * Time: 12:39:21 PM
- */
 public enum PasswordRule
```
Comment (labkey-matthewb) [inline api/src/org/labkey/api/security/PasswordRule.java:35]: Change to class or record to anticipate user defined rules?

File: api/src/org/labkey/api/security/PasswordRule.java [hunk 3]
```diff
@@ -216,18 +296,68 @@ public boolean isValidToStore(String password1, String password2, User user, Col
...
+     * @return HtmlString providing a summary of the rules
+     */
+    @NotNull
+    public HtmlString getSummaryRuleHtml()
+    {
+        return _summaryRuleHtml;
+    }
 
     // 100 most-used passwords on RockYou
     private static final String weakPasswords = "123456\n"+
```
Comment (labkey-matthewb) [inline api/src/org/labkey/api/security/PasswordRule.java:365]: Make actual JunitTest

File: api/src/org/labkey/api/security/PasswordRule.java [hunk 4]
```diff
@@ -216,18 +296,68 @@ public boolean isValidToStore(String password1, String password2, User user, Col
...
+
+    protected String getMinimumLengthText()
+    {
+        return StringUtilsLabKey.spellOut(getMinimumLength());
+    }
+
+    protected boolean containsPersonalInfo(@NotNull String password, @NotNull User user, @Nullable Collection<String> messages)
+    {
+        return
+            containsPersonalInfo(password, user.getEmail(), "email address", messages) &&
```
Comment (labkey-matthewb) [inline api/src/org/labkey/api/security/PasswordRule.java]: ||

**Comment by labkey-matthewb** [comment]
Would be nice to move "passwords do not match" check to client form.

**Review by labkey-matthewb — CHANGES_REQUESTED** [review]

File: api/src/org/labkey/api/security/PasswordRule.java (hunk 4 above)
Comment (labkey-adam) [reply api/src/org/labkey/api/security/PasswordRule.java]: Oops... thanks!

File: api/src/org/labkey/api/security/PasswordRule.java (hunk 1 above)
Comment (labkey-adam) [reply api/src/org/labkey/api/security/PasswordRule.java]: Fixed

**Comment by labkey-adam** [comment]
> Would be nice to move "passwords do not match" check to client form.

Agreed. Unfortunately, I'd have to update MacCoss SignUpController as well.

**Review by labkey-matthewb — APPROVED** [review]

**Pull request merged**